### PR TITLE
Energy output

### PR DESCRIPTION
**Pull request by krenzland**
This PR develops the energy output developed in #494 into a "proper" writer, with csv output and sync point support.

Additionally, it computes the wave field energy for elastic and acoustic (including gravitational free surface) media.

I've evaluated this setup by comparing the energy computed in SeisSol with energy computed by integrating the analytical solution numerically.
I've used two scenarios:

- Water tank / ocean scenario, includes acoustic, acoustic kinetic and gravitational energy
-  Snell's law, includes elastic, elastic kinetic, acoustic and acoustic kinetic energy but is basically a 2D solution.

It should work for 3D solutions as well, given that I use the correct indices - would be great if someone could double check that.

### CODE REVIEW

File: Documentation/energy-output.rst [hunk 1]
```diff
@@ -0,0 +1,53 @@
+Energy output
+============
+
+Introduction
+------------
+
+The energy output computes the energy of the simulation. It is divided into multiple parts:
+
+- Energy in the water layer (= acoustic medium).
+    - Gravitational energy
```
Comment (Thomas-Ulrich) [inline Documentation/energy-output.rst:10]: would be nice to have the formula of each of these energies in the doc...

Comment (Thomas-Ulrich) [reply Documentation/energy-output.rst:10]: I can provide the code for the energies that I added.

Comment (krenzland) [reply Documentation/energy-output.rst:10]: Yes, good idea!

File: src/ResultWriter/EnergyOutput.cpp [hunk 2]
```diff
@@ -98,57 +167,243 @@ void seissol::writer::printEnergies(GlobalData const* global,
...
+      << "acoustic_kinetic_energy,"
+      << "elastic_energy,"
+      << "elastic_kinetic_energy,"
+      << "total_frictional_work,"
+      << "static_frictional_work,"
+      << "plastic_moment" << std::endl;
+}
+
+void EnergyOutput::writeEnergies(double time) {
+  const auto rank = MPI::mpi.rank();
```
Comment (Thomas-Ulrich) [inline src/ResultWriter/EnergyOutput.cpp:401]: not needed?

File: src/ResultWriter/EnergyOutput.cpp [hunk 3]
```diff
@@ -98,57 +167,243 @@ void seissol::writer::printEnergies(GlobalData const* global,
...
+        const auto p = numSub(qp, pIdx);
+
+        const double curAcousticEnergy = (p * p) / ( 2 * K);
+        totalAcousticEnergyLocal += curWeight * curAcousticEnergy;
+        totalAcousticKineticEnergyLocal += curWeight * curKineticEnergy;
+      } else {
+        // Elastic
+        totalElasticKineticEnergyLocal += curWeight * curKineticEnergy;
+        auto getStressIndex = [](int i, int j) {
+          auto lookup = std::array<std::array<int, 3>, 3>{{{0, 3, 5}, {3, 1, 4}, {5, 4, 2}}};
```
Comment (Thomas-Ulrich) [inline src/ResultWriter/EnergyOutput.cpp:267]: Would it be worth initializing this array outside the function and the loops and passing it as a reference?

File: src/ResultWriter/EnergyOutput.cpp [hunk 4]
```diff
@@ -98,57 +167,243 @@ void seissol::writer::printEnergies(GlobalData const* global,
...
+  std::vector<Vertex> const& vertices = meshReader->getVertices();
+
+  const auto g = SeisSol::main.getGravitationSetup().acceleration;
+
+  // Note: Default(none) is not possible, clang requires data sharing attribute for g, gcc forbids it
+#ifdef _OPENMP
+#pragma omp parallel for schedule(static) reduction(+ : totalGravitationalEnergyLocal, totalAcousticEnergyLocal, totalAcousticKineticEnergyLocal, totalElasticEnergyLocal, totalElasticKineticEnergyLocal) shared(elements, vertices, lts, ltsLut, global)
+#endif
+  for (std::size_t elementId = 0; elementId < elements.size(); ++elementId) {
+#ifdef USE_ELASTIC
```
Comment (Thomas-Ulrich) [inline src/ResultWriter/EnergyOutput.cpp:207]: would the code not work also for viscoelastic2?

File: src/ResultWriter/EnergyOutput.cpp [hunk 5]
```diff
@@ -98,57 +167,243 @@ void seissol::writer::printEnergies(GlobalData const* global,
...
+          return lookup[i][j];
+        };
+        auto getStress = [&](int i, int j) { return numSub(qp, getStressIndex(i, j)); };
+
+        const auto lambda = material.local.lambda;
+        const auto mu = material.local.mu;
+        const auto dilation = getStress(0, 0) + getStress(1, 1) + getStress(2, 2);
+        auto computeStrain = [&](int i, int j) {
+          double strain = 0.0;
+          const auto factor = -1.0 * (lambda) / (2 * mu * (3 * lambda + 2 * mu));
```
Comment (Thomas-Ulrich) [inline src/ResultWriter/EnergyOutput.cpp:277]: 2.0 3.0 2.0?

File: src/ResultWriter/EnergyOutput.cpp [hunk 6]
```diff
@@ -98,57 +167,243 @@ void seissol::writer::printEnergies(GlobalData const* global,
...
+        totalElasticKineticEnergyLocal += curWeight * curKineticEnergy;
+        auto getStressIndex = [](int i, int j) {
+          auto lookup = std::array<std::array<int, 3>, 3>{{{0, 3, 5}, {3, 1, 4}, {5, 4, 2}}};
+          return lookup[i][j];
+        };
+        auto getStress = [&](int i, int j) { return numSub(qp, getStressIndex(i, j)); };
+
+        const auto lambda = material.local.lambda;
+        const auto mu = material.local.mu;
+        const auto dilation = getStress(0, 0) + getStress(1, 1) + getStress(2, 2);
```
Comment (Thomas-Ulrich) [inline src/ResultWriter/EnergyOutput.cpp:274]: not dilatation (strain), rather sumUniaxialStresses

File: src/ResultWriter/EnergyOutput.cpp [hunk 7]
```diff
@@ -98,57 +167,243 @@ void seissol::writer::printEnergies(GlobalData const* global,
...
+      static_assert(numQuadraturePointsTri ==
+                    init::rotatedFaceDisplacementAtQuadratureNodes::Shape[0]);
+      auto rotatedFaceDisplacement =
+          init::rotatedFaceDisplacementAtQuadratureNodes::view::create(displQuadData.data());
+      for (unsigned i = 0; i < rotatedFaceDisplacement.shape(0); ++i) {
+        // See for example (Saito, Tsunami generation and propagation, 2019) section 3.2.3 for
+        // derivation.
+        const auto displ = rotatedFaceDisplacement(i, 0);
+        const auto curEnergy = 0.5 * rho * g * displ * displ;
+        const auto curWeight = 2 * surface * quadratureWeightsTri[i];
```
Comment (Thomas-Ulrich) [inline src/ResultWriter/EnergyOutput.cpp:336]: 2.0?

File: src/ResultWriter/EnergyOutput.cpp [hunk 8]
```diff
@@ -98,57 +167,243 @@ void seissol::writer::printEnergies(GlobalData const* global,
...
+  // Note: Default(none) is not possible, clang requires data sharing attribute for g, gcc forbids it
+#ifdef _OPENMP
+#pragma omp parallel for schedule(static) reduction(+ : totalGravitationalEnergyLocal, totalAcousticEnergyLocal, totalAcousticKineticEnergyLocal, totalElasticEnergyLocal, totalElasticKineticEnergyLocal) shared(elements, vertices, lts, ltsLut, global)
+#endif
+  for (std::size_t elementId = 0; elementId < elements.size(); ++elementId) {
+#ifdef USE_ELASTIC
+    real volume = MeshTools::volume(elements[elementId], vertices);
+    CellMaterialData& material = ltsLut->lookup(lts->material, elementId);
+    auto& cellInformation = ltsLut->lookup(lts->cellInformation, elementId);
+    auto& faceDisplacements = ltsLut->lookup(lts->faceDisplacements, elementId);
```
Comment (Thomas-Ulrich) [inline src/ResultWriter/EnergyOutput.cpp:211]: so faceDisplacements are stored in a face-aligned referential, and we want to rotate it to the global cartesian referential?

**Review by Thomas-Ulrich — CHANGES_REQUESTED** [review]
Nice code and new features!

**Comment by Thomas-Ulrich** [comment]
Ps: Maybe there is nothing supporting keeping the terminal output of the energy output.

File: generated_code/aderdg.py [hunk 9]
```diff
@@ -96,6 +96,18 @@ def __init__(self, order, multipleSimulations, matricesDir):
...
+    self.db.update(
+      parseXMLMatrixFile("{}/nodal/gravitational_energy_matrices_{}.xml".format(
+        matricesDir,
+        self.order,
+        alignStride=self.alignStride
+      ))
+    )
+
+
+    self.V2nTo2JacobiQuad = tensor_from_constant_expression('V2nTo2JacobiQuad', self.db.V2mTo2JacobiQuad['ik'] * \
```
[the remark this answers is not in the document]
Comment (krenzland) [reply generated_code/aderdg.py:108]: Please don't ask...
The naming is a mess. M is the mass matrix. But the matrix with the prefix M is the one without the mass matrix, the one without prefix M is with mass matrix.
We'll change that but the matrices were created by some dubious Maple scripts. (We do not have a Maple license).
We're working on this. :)

File: src/ResultWriter/EnergyOutput.cpp (hunk 4 above)
Comment (krenzland) [reply src/ResultWriter/EnergyOutput.cpp:207]: Should work. But energy lost due to attenuation is not tracked of course. (Same as energy lost due to outflow bc)

Comment (Thomas-Ulrich) [reply src/ResultWriter/EnergyOutput.cpp]: next thing to do :-)

File: src/ResultWriter/EnergyOutput.cpp [hunk 10]
```diff
@@ -77,17 +153,10 @@ real seissol::writer::computeStaticWork(GlobalData const* global,
...
-                                    seissol::initializers::LTS* i_lts,
-                                    seissol::initializers::Lut* i_ltsLut,
-                                    bool usePlasticity) {
-  double totalFrictionalWorkLocal = 0.0;
-  double staticFrictionalWorkLocal = 0.0;
-  double plasticMomentLocal = 0.0;
+void EnergyOutput::computeDynamicRuptureEnergies() {
+  double& totalFrictionalWork = energiesStorage.totalFrictionalWork();
+  double& staticFrictionalWork = energiesStorage.staticFrictionalWork();
+  double& plasticMoment = energiesStorage.plasticMoment();
```
Comment (Thomas-Ulrich) [inline src/ResultWriter/EnergyOutput.cpp:159]: plasticMoment is not a dynamicRuptureEnergy

File: src/ResultWriter/EnergyOutput.cpp [hunk 11]
```diff
@@ -223,9 +199,9 @@ void EnergyOutput::computeEnergies() {
 #pragma omp parallel for schedule(static) reduction(+ : totalGravitationalEnergyLocal, totalAcousticEnergyLocal, totalAcousticKineticEnergyLocal, totalElasticEnergyLocal, totalElasticKineticEnergyLocal) shared(elements, vertices, lts, ltsLut, global)
```
Comment (Thomas-Ulrich) [inline src/ResultWriter/EnergyOutput.cpp:199]: reduction(+ :  totalPlasticMoment))

**Review by Thomas-Ulrich — CHANGES_REQUESTED** [review]
there are still comments not answered, right?

File: src/ResultWriter/EnergyOutput.cpp [hunk 12]
```diff
@@ -98,57 +167,243 @@ void seissol::writer::printEnergies(GlobalData const* global,
...
-  staticFrictionalWorkGlobal = staticFrictionalWorkLocal;
-  plasticMomentGlobal = plasticMomentLocal;
+    auto numSub = numericalSolution;
+#endif
+    for (size_t qp = 0; qp < numQuadraturePointsTet; ++qp) {
+      constexpr int uIdx = 6;
+      const auto curWeight = jacobiDet * quadratureWeightsTet[qp];
+      const auto rho = material.local.rho;
+
+      const auto u = numSub(qp, uIdx + 0);
```
[the remark this answers is not in the document]
Comment (krenzland) [reply src/ResultWriter/EnergyOutput.cpp:244]: The code is copied from resultwriter. It should work, but is untested currently. It's currently broken, see https://github.com/SeisSol/SeisSol/pull/385

File: src/ResultWriter/EnergyOutput.cpp (hunk 3 above)
Comment (krenzland) [reply src/ResultWriter/EnergyOutput.cpp:262]: The compiler should recognize that it is constant. I'll mark it static const to enforce this

File: generated_code/matrices/nodal/gravitational_energy_matrices_2.xml [hunk 13]
```diff
@@ -0,0 +1,33 @@
+<?xml version="1.0" ?>
```
[the remark this answers is not in the document]
Comment (krenzland) [reply generated_code/matrices/nodal/gravitational_energy_matrices_2.xml:1]: https://github.com/sebwolf-de/SeisSol-Matrices/blob/master/seissol_matrices/energy.py

File: src/ResultWriter/EnergyOutput.cpp [hunk 14]
```diff
@@ -98,57 +167,243 @@ void seissol::writer::printEnergies(GlobalData const* global,
...
+    double quadratureWeightsTri[numQuadraturePointsTri];
+    seissol::quadrature::TriangleQuadrature(
+        quadraturePointsTri, quadratureWeightsTri, quadPolyDegree);
+
+    // Needed to weight the integral.
+    const auto jacobiDet = 6 * volume;
+
+    alignas(ALIGNMENT) real numericalSolutionData[tensor::dofsQP::size()];
+    auto numericalSolution = init::dofsQP::view::create(numericalSolutionData);
+    // Evaluate numerical solution at quad. nodes
```
[the remark this answers is not in the document]
Comment (krenzland) [reply src/ResultWriter/EnergyOutput.cpp:227]: It works in the same way as in ResultWriter. The quadrature points are hardcoded in the matrix. It's just a polynomial evaluation. Yes, the size is numQuadraturePointsTet (+ potentially alignment)

**Review by Thomas-Ulrich — APPROVED** [review]

**Pull request merged**